### PR TITLE
Implement controlled upgrades

### PR DESCRIPTION
**Pull request by mateoflorido**
## Summary
This pull request implements the controlled upgrade process for the Kubernetes Worker charm.
## Changes
- Implement the `upgrade` action.
## Related Pull Requests
- Depends on: https://github.com/charmed-kubernetes/charm-lib-kubernetes-snaps/pull/8

### CODE REVIEW

File: src/charm.py [hunk 1]
```diff
@@ -289,6 +291,30 @@ def _install_cni_binaries(self):
...
+                log.exception(error_message)
+                status.add("Snap upgrade failed. Check action results for more information.")
+                event.fail(error_message)
+            else:
+                result_message = (
+                    f"Successfully upgraded Kubernetes snaps to the '{channel}' channel."
+                )
+                log.info(result_message)
+                event.set_results({"result": result_message})
+
```
Comment (addyess) [inline src/charm.py:317]: this is same handler in both charms except for the `control_plane=True` isn't it?  

any chance you'd like to bury the handler in the `kubernetes_snaps` library?

Comment (addyess) [reply src/charm.py:317]: the `event` should have a link to the framework -- and the framework should have some metadata.  and that metadata should know the `charm-name` (not application name)

Comment (addyess) [reply src/charm.py:317]: or -- you could write a SMALL handler in each charm that passes the right flag to the library handler

Comment (mateoflorido) [reply src/charm.py:317]: Is it possible to retrieve the current unit and the `channel` config value from the event? I considered this approach, but then the afterthought made me reconsider...

Comment (addyess) [reply src/charm.py:317]: so the event has in it the [`framework`](https://ops.readthedocs.io/en/latest/#ops.Framework) which exposes:
* `meta` -- metadata about the charm
* `model` 
   * model has the `config` property
   * model has the `unit` property

**Review by addyess — APPROVED** [review]
LGTM

**Pull request merged**